packages/workers-shared/asset-worker/src/handler.ts

```diff
@@ -0,0 +1,589 @@
+import {
+	InternalServerErrorResponse,
+	MethodNotAllowedResponse,
+	NotFoundResponse,
+	NotModifiedResponse,
+	OkResponse,
+	TemporaryRedirectResponse,
+} from "./responses";
+import { getHeaders } from "./utils/headers";
+import type { AssetConfig } from "../../utils/types";
+import type EntrypointType from "./index";
+
+export const handleRequest = async (
+	request: Request,
+	configuration: Required<AssetConfig>,
+	// HELP: is this a legit pattern? can i do this?? feels wrong
+	exists: typeof EntrypointType.prototype.exists,
+	getByETag: typeof EntrypointType.prototype.getByETag
+) => {
+	const { pathname, search } = new URL(request.url);
+
+	const intent = await getIntent(pathname, configuration, exists);
+	if (!intent) {
+		return new NotFoundResponse();
+	}
+
+	// if there was a POST etc. to a route without an asset
+	// this should be passed onto a user worker if one exists
+	// so prioritise returning a 404 over 405?
+	const method = request.method.toUpperCase();
+	if (!["GET", "HEAD"].includes(method)) {
+		return new MethodNotAllowedResponse();
+	}
+	if (intent.redirect) {
+		return new TemporaryRedirectResponse(intent.redirect + search);
+	}
+	if (!intent.asset) {
+		return new InternalServerErrorResponse(new Error("Unknown action"));
+	}
+
+	const asset = await getByETag(intent.asset.eTag);
+
+	const headers = getHeaders(intent.asset.eTag, asset.contentType, request);
+
+	const strongETag = `"${intent.asset.eTag}"`;
+	const weakETag = `W/${strongETag}`;
+	const ifNoneMatch = request.headers.get("If-None-Match") || "";
+	if ([weakETag, strongETag].includes(ifNoneMatch)) {
+		return new NotModifiedResponse(null, { headers });
+	}
+
+	const body = method === "HEAD" ? null : asset.readableStream;
+	switch (intent.asset.status) {
+		case 404:
+			return new NotFoundResponse(body, { headers });
+		case 200:
+			return new OkResponse(body, { headers });
+	}
+};
+
+type Intent =
+	| {
+			asset: { eTag: string; status: 200 | 404 };
+			redirect: null;
+	  }
+	| { asset: null; redirect: string }
+	| null;
+
+export const getIntent = async (
+	pathname: string,
+	configuration: Required<AssetConfig>,
+	exists: typeof EntrypointType.prototype.exists,
+	skipRedirects = false
+): Promise<Intent> => {
+	switch (configuration.html_handling) {
+		case "auto-trailing-slash": {
+			return htmlHandlingAutoTrailingSlash(
+				pathname,
+				configuration,
+				exists,
+				skipRedirects
+			);
+		}
+		case "force-trailing-slash": {
+			return htmlHandlingForceTrailingSlash(
+				pathname,
+				configuration,
+				exists,
+				skipRedirects
+			);
+		}
+		case "drop-trailing-slash": {
+			return htmlHandlingDropTrailingSlash(
+				pathname,
+				configuration,
+				exists,
+				skipRedirects
+			);
+		}
+		case "none": {
+			return htmlHandlingNone(pathname, configuration, exists);
+		}
+	}
+	return null;
+};
+
+const htmlHandlingAutoTrailingSlash = async (
```

Comment:
These different handling functions are really complex and contain a lot of duplicated logic. Not a blocker for merge, but it would be good to explore ways to make this a bit more readable